core/src/main/java/org/apache/iceberg/avro/AvroSchemaWithTypeVisitor.java

```diff
@@ -97,17 +100,89 @@ private static <T> T visitUnion(Type type, Schema union, AvroSchemaWithTypeVisit
+      } else {
+        boolean relatedFieldInStructFound = false;
+        Types.StructType struct = type.asStructType();
+        if (fieldIndexInStruct < struct.fields().size() &&
+                UNION_TAG_FIELD_NAME.equals(struct.fields().get(fieldIndexInStruct).name())) {
+          fieldIndexInStruct++;
+        }
+
+        if (fieldIndexInStruct < struct.fields().size()) {
+          // If a NULL type is found before current type, the type index is one larger than the actual type index which
+          // can be used to track the corresponding field in the struct of Iceberg schema.
+          int actualTypeIndex = nullTypeFound ? typeIndex - 1 : typeIndex;
+          String structFieldName = type.asStructType().fields().get(fieldIndexInStruct).name();
+          int indexFromStructFieldName = Integer.valueOf(structFieldName.substring(5));
+          if (actualTypeIndex == indexFromStructFieldName) {
+            relatedFieldInStructFound = true;
+            options.add(visit(type.asStructType().fields().get(fieldIndexInStruct).type(), schema, visitor));
+            fieldIndexInStruct++;
+          }
+        }
+
+        if (!relatedFieldInStructFound) {
+          visitNotProjectedTypeInComplexUnion(schema, visitor, options);
         }
       }
+      typeIndex++;
     }
-    return visitor.union(type, union, options);
+  }
+
+  // If a field is not projected, a corresponding field in the struct of Iceberg schema cannot be found
+  // for current type of union in Avro schema, a reader for current type still needs to be created and
+  // used to make the reading of Avro file successfully. In this case, an pseudo Iceberg type is converted from
+  // the Avro schema and is used to create the option for the reader of the current type which still can
+  // read the corresponding content in Avro file successfully.
+  private static <T> void visitNotProjectedTypeInComplexUnion(Schema schema,
+                                                                            AvroSchemaWithTypeVisitor<T> visitor,
+                                                                            List<T> options) {
+    Type iType = AvroSchemaUtil.convert(schema);
+    if (schema.getType().equals(Schema.Type.RECORD)) {
+      // When the type of Avro schema is RECORD, the fields under it must have the property of "field-id".
+      // However, the "field-id" is not set in previous steps as the corresponding Iceberg type is not projected
+      // and no field id can be found for this field in Iceberg schema.
+      // Therefore, a name mapping is created based on the Avro schema and its corresponding Iceberg type.
+      // The field-id from the resulted name mapping is assigned as the property of "field-id" of each
+      // field under the Avro schema.
+      NameMappingWithAvroSchema nameMappingWithAvroSchema = new NameMappingWithAvroSchema();
+      MappedFields nameMapping = AvroWithPartnerByStructureVisitor.visit(
```

Comment:
do field-ids need to be unique across all the (top level & nested) fields in the schema? if yes, does this visitor achieve that?

Reply:
yes